vars/runATH.groovy

```diff
@@ -0,0 +1,217 @@
+
+            // ATH
+            if (isLocalATH) { // Deal with already existing ATH sources
+                athSourcesFolder = athUrl - "file://"
+            } else {
+                dir(athSourcesFolder) {
+                    checkout changelog: true, poll: false, scm: [$class           : 'GitSCM', branches: [[name: athRevision]],
+                                                                 userRemoteConfigs: [[url: athUrl]]]
+                }
+            }
+            dir(athSourcesFolder) {
+                // We may need to run things in parallel later, so avoid several checkouts
+                stash name: "athSources"
+            }
+
+            // Jenkins war
+            if (isVersionNumber) {
+                def downloadCommand = "mvn dependency:copy -Dartifact=org.jenkins-ci.main:jenkins-war:${jenkins}:war -DoutputDirectory=. -Dmdep.stripVersion=true"
+                dir("deps") {
+                    if (infra.isRunningOnJenkinsInfra()) {
+                        def settingsXml = "${pwd tmp: true}/settings-azure.xml"
+                        writeFile file: settingsXml, text: libraryResource('settings-azure.xml')
+                        downloadCommand = downloadCommand + " -s ${settingsXml}"
+                    }
+                    withEnv(env) {
+                        sh downloadCommand
+                    }
+                    sh "cp jenkins-war.war jenkins.war"
+                    stash includes: 'jenkins.war', name: 'jenkinsWar'
+                }
+            } else {
+                sh("curl -o jenkins.war -L ${jenkinsURl}")
+                stash includes: '*.war', name: 'jenkinsWar'
+            }
+
+        }
+    })
+
+    ensureInNode(env, env.RUN_ATH_DOCKER_NODE ?: 'docker && highmem', {
+        if (skipExecution) {
+            return
+        }
+        stage("Running ATH") {
+            dir("athSources") {
+                unstash name: "athSources"
+                def uid = sh(script: "id -u", returnStdout: true)
+                def gid = sh(script: "id -g", returnStdout: true)
+                athContainerImage = docker.build('jenkins/ath', "--build-arg=uid='$uid' --build-arg=gid='$gid' -f src/main/resources/ath-container/Dockerfile .")
+            }
+
+            def testsToRun = metadata.tests?.join(",")
+            def categoriesToRun = metadata.categories?.join(",")
+            def browsers = metadata.browsers ?: ["firefox"]
+            def failFast = metadata.failFast ?: false
+            def rerunCount = metadata.rerunFailingTestsCount ?: 0
+            def localSnapshots = metadata.useLocalSnapshots ?: true
+
+            if (testsToRun == null && categoriesToRun == null) {
+                categoriesToRun = defaultCategory
+            }
+
+            def testingbranches = ["failFast": failFast]
+            for (browser in browsers) {
+                if (supportedBrowsers.contains(browser)) {
+
+                    def currentBrowser = browser
```

Comment:
Doesn't matter much, but why is this brought out into a separate variable?  Looks like it's only used for reads and never reassigned.

Reply:
As I am defining closures instead of running code the looping over `browser` mutated the contents of the closure, resulting in having all the closures with strings interpolated to the last value of `browser`